### DISCUSSION TITLE
Why my .dll can't use in winui3, but it can use at wpf?

### DISCUSSION THREAD

**Discussion by xtuzy** [microsoft/WindowsAppSDK#2532] [Q&A]
Hello, i have a .dll file, it compiled from c++, i set it at a netstandard class library, and use p/invoke to load method of .dll.
When i used it at wpf, it run well, now i want use winui3 as ui , so i add it to winui3 for desktop project, but when i run method of .dll, app will throw exception.
Have any suggestions?

**Comment by huoyaoyuan** [comment]
We can't provide any help without knowing what exception is thrown.
Since there is managed/unmanaged interop, I'd suggest to check bitness of the dll and the exe.

**Reply by xtuzy** [reply]
> We can't provide any help without knowing what exception is thrown. Since there is managed/unmanaged interop, I'd suggest to check bitness of the dll and the exe.

I only get this.
```
Unhandled exception at 0x00007FFE3507E50D (Microsoft.ui.xaml.dll) in WinUI3App.exe: 0xC000027B: An application-internal exception has occurred (parameters: 0x0000018EF94FFDA0, 0x0000000000000002).
```
App run to this method that init .dll library will throw exception
![sharex-20220519142608](https://user-images.githubusercontent.com/17793881/169225000-0c126185-bae0-4351-ae01-7f734295aa3b.png)

[Source at here](https://github.com/xtuzy/ApplePDF/tree/ApplePDF)

**Reply by huoyaoyuan** [reply]
It looks like invocation chain is WinUI(unmanaged) -> your page/window(managed) -> pdfium(unmanaged).
An exception was thrown from managed code, then caught by unmanaged WinUI and the detail was lost.
Can you add a try around the managed call and see what's thrown?

**Reply by xtuzy** [reply]
I add a try:
```
void init()
{
    try
    {
        fpdfview.FPDF_InitLibrary();
    }
    catch (Exception ex)
    {
        throw ex;
    }
}
```
Get ex: "Unable to load DLL 'pdfium' or one of its dependencies: The specified module could not be found. (0x8007007E)"
![image](https://user-images.githubusercontent.com/17793881/169308999-c25566d4-03c9-48c9-a541-267b3a3525be.png)

Is it mean pdfium.dll use something that not exit at winui3 ?

**Reply by huoyaoyuan** [reply]
I looked through your code and I suspect that your way loading native library is problematic. In this case your pdfium.dll is probably lost.

I've not successfully make native library work with `ProjectReference` ever.

**Reply by xtuzy** [reply]
Thank you point the key problem, i try to move x64 dll to app root directory, it work! 
But do you have any advice to let ProjectReference work? i have many dll at runtimes folder, i found this folder at class project  will be copy to App when builded, but App can't know to load from this folder.
Another, i found Skiasharp can set correct dll to root folder by nuget.
![image](https://user-images.githubusercontent.com/17793881/169436417-6c3d341c-b907-492c-a34a-d5914474dcf4.png)
